[lhy-hoyin] iP

lhy-hoyin opened the pull request:
# Duke
> *Sorry, no quote available :<*

Duke <sub>copy</sub> is a simple task tracker. It's quick and responsive, easy-to-use, and **FREE**.
Developed by: @lhy-hoyin 😎 

[Website](https://lhy-hoyin.github.io/ip/)

## Features
1. List all tasks
2. Add a task
    - To-do task
    - Deadline task
    - Event task
6. Mark task(s) as completed or uncompleted
7. Delete task(s)
8. Search for task(s) by providing some words found in the task description

### Extra Features Ideas (may ~not~ be implemented)
> Don't be too hopefully 🙃 
* Option to `list` all uncompleted (or completed) tasks
* Manage multiple task lists
* Have a config file to allow user preferences
  * Custom date/time input format
  * Custom separator for mass operations (e.g. mark, unmark, delete)
* Background auto-save


## Work In Progress (_hopefully coming soon_)
- [X] README
- [ ] Add more `Date` and `Time` parsing formats (allow custom formats?)
- [X] Include more JUnit tests
- [ ] Automate more using Gradle
  - [X] Run checkstyle after building
- [ ] Add `help` for in-application support (see [here](https://nus-cs2103-ay2223s2.github.io/website/se-book-adapted/projectDuke/index.html#c-help))
- [ ] Add more Flags to ...
  - [ ] load configuration files
  - [ ] open specific task list (requires multi- task lists feature)


## Bugs
Known/Potential Bugs
- [X] Printing `Time`  may not be 4-digits
- [X] `mark`/`unmark`/`delete` might refer to the wrong index
  -  happens when non-actual list index is printed 
  - e.g. `find` prints index according to `find` results, not actual index of the task in the list
- [X] *FATAL*: Catch exceptions when parsing date/time (currently assume user always enter in correct format)
- [X] Typing `exit` (or other keywords) into the GUI does not trigger the exiting behavior
- [ ] When GUI text becomes too long, the label text are `...` (including new text)

Find a bug? Report it [here](https://github.com/lhy-hoyin/ip/issues).


## Extra - Code References for Reuse
Mentioned in [[iP] Level-7 file pathing #31](https://github.com/nus-cs2103-AY2223S2/forum/issues/31)
```java
private void writeToFile(String filePath, String fileContent) {
    try {
        Path f = Paths.get(filePath);
        Files.createDirectories(f.getParent()); // Create directory (if not exist)
        if (!Files.exists(f)) {
            Files.createFile(f); // Create non-existing file
        }
        Files.writeString(f, fileContent); // Write to file
    } catch (IOException e) {
        throw new RuntimeException(e);
    }
}
```

Commit:
Add Gradle support

lhy-hoyin committed:
Implemented skeletal working version to start, run (echo user input), and exit program.

lhy-hoyin committed:
Added chatbot customisation

lhy-hoyin committed:
Fixed bug where program does not truly exits upon saying "bye".
Extracted out command detection. Added new UNKNOWN state.

lhy-hoyin committed:
Added support for various exit keywords.
Adjusted program loop exiting.

lhy-hoyin committed:
Implemented list feature (detect state, add to list, read from list).
Added LIST state, Changed EHCO state to ADD State.
Enhanced state detection (lower first, first word only).

lhy-hoyin committed:
Enforced requirement of a "add" keyword for adding to list.

lhy-hoyin committed:
Converted ArrayList<String> to ArrayList<Task>

lhy-hoyin committed:
Added functionality for marking and marking task

lhy-hoyin committed:
Optimised & merged mark and unmark code. Added comments for future exceptions to watch for.

lhy-hoyin committed:
Implemented subclasses of Task (Todo, Deadline, Event)

lhy-hoyin committed:
Updated "add" to "todo"

lhy-hoyin committed:
Updated UI message for todo task, extracted certain print message into functions.

lhy-hoyin committed:
Implemented support for deadline- and event- typed tasks

lhy-hoyin committed:
minor changes (added comments, moved State into `Duke` class)

lhy-hoyin committed:
Added automated testing for UI (note, user input is not part of expected output).
Had to clean up ending space of the logo

lhy-hoyin committed:
Added DukeException class

lhy-hoyin committed:
Implemented code to handle various exceptions.

lhy-hoyin committed:
Implemented functionality to delete task

lhy-hoyin committed:
Made Duke static methods private.
Implemented default case for program loop.

lhy-hoyin committed:
Update EXPECTED.txt

lhy-hoyin committed:
Move some Duke related methods from static context into a instance
context

lhy-hoyin committed:
Implement basic writing of data (in CSV format) to data file
upon task creation and program exit

To do:
 * implement reading from data file during program startup
 * save to file during mark/unmark/delete
 * extra: add a "save" command to allow user to save manually

lhy-hoyin committed:
Update .gitignore to exclude data file (task.csv)

lhy-hoyin committed:
Implement data save upon mark/unmark/delete

Add "save" keyword to manually do data save

lhy-hoyin committed:
Implement loading of data from file
Overload constructors for Task (and children classes) for isDone
Change the file save data for isDone attribute

lhy-hoyin committed:
Implement parsing to Local Date/Time/DateTime
for specific String format

lhy-hoyin committed:
Provide alternative implementation for parsing date and time
using built-in parsing

lhy-hoyin committed:
Update Deadline and Event class to store in LocalDateTime

lhy-hoyin committed:
Merge branch 'branch-Level-7'

* branch-Level-7:
  Implement loading of data from file Overload constructors for Task (and children classes) for isDone Change the file save data for isDone attribute
  Implement data save upon mark/unmark/delete
  Update .gitignore to exclude data file (task.csv)
  Implement basic writing of data (in CSV format) to data file upon task creation and program exit

lhy-hoyin committed:
Merge branch 'branch-Level-8'

* branch-Level-8:
  Update Deadline and Event class to store in LocalDateTime
  Provide alternative implementation for parsing date and time
    using built-in parsing
  Implement parsing to Local Date/Time/DateTime
    for specific String format

# Conflicts:
#	src/main/java/Deadline.java
#	src/main/java/Duke.java
#	src/main/java/Event.java

Merge conflicts resolved

lhy-hoyin committed:
Organize the classes into suitable java packages (minimal)

lhy-hoyin committed:
Refactor code to separate into more OOP classes
Move storage related into its own package

lhy-hoyin committed:
Convert detectState(..) to fully use switch case

lhy-hoyin committed:
Remove enum State
Merge detectState(..) switch case into main program loop

lhy-hoyin committed:
Delegate task parameters parsing to TaskList when adding task

lhy-hoyin committed:
no message

lhy-hoyin committed:
Remove getInner() function that violates OOP concepts

lhy-hoyin committed:
Separate writeToFile functionality from saveDataToFile()

lhy-hoyin committed:
Reverse commit #340a113

lhy-hoyin committed:
Make duke components final

lhy-hoyin committed:
Add parsing date/time with arbitrary separator

lhy-hoyin committed:
Merge remote-tracking branch 'origin/add-gradle-support'

lhy-hoyin committed:
Use gradle to build and run Duke (Minimal)

lhy-hoyin committed:
Implement find function to look for a task

lhy-hoyin committed:
Added JavaDocs for all classes and most methods

lhy-hoyin committed:
Adds bounds checking when removing Task from TaskList
Changes some access modifiers
Renames displayLine() printBufferLine()
Reorders import statements in Parser.java

lhy-hoyin committed:
Conforms code to coding standards

lhy-hoyin committed:
Merge branch 'branch-Level-9'

* branch-Level-9:
  Implement find function to look for a task

lhy-hoyin committed:
Adds punctuation behind each parameter description

lhy-hoyin committed:
Merge branch 'branch-A-JavaDoc'

# Conflicts:
#	src/main/java/duke/TaskList.java
#	src/main/java/duke/Ui.java
#	src/main/java/duke/storage/Storable.java

lhy-hoyin committed:
Merge branch 'branch-A-CodingStandard'

* branch-A-CodingStandard:
  Conforms code to coding standards

# Conflicts:
#	src/main/java/duke/Parser.java

Resolve conficts

lhy-hoyin committed:
Add JUnit Test for Parser and Ui

lhy-hoyin committed:
Use Checkstyle with Gradle
Checkstyle configuration from AddressBook Level 3 project

lhy-hoyin committed:
Enforce coding standards flagged by checkstyle

lhy-hoyin committed:
Merge branch 'branch-A-CodingStandard'

* branch-A-CodingStandard:
  Enforce coding standards flagged by checkstyle

lhy-hoyin committed:
Sets up multi-mode launching.
Can launch in CLI mode (existing) or GUI mode (to be implemented).

RyanQiu1 commented on src/test/java/duke/ParserTest.java:
```diff
@@ -0,0 +1,38 @@
+    private Parser parser = new Parser();
+
+    @Test
+    void parseDateTime_spaceSeperator_success() {
```
Perhaps a shorter and intuitive method name here?

lhy-hoyin replied:
I'm trying to follow the `whatIsBeingTested_descriptionOfTestInputs_expectedOutcome` for test methods as mentioned [here](https://nus-cs2103-ay2223s2.github.io/website/schedule/week3/topics.html#c-to-java-junit-junit-basic).

RyanQiu1 commented on src/test/java/duke/ParserTest.java:
```diff
@@ -0,0 +1,38 @@
+    }
+
+    @Test
+    void parseDateTime_arbitrarySeperator_success() {
```
Perhaps a shorter and intuitive method name here?

lhy-hoyin replied:
I'm trying to follow the `whatIsBeingTested_descriptionOfTestInputs_expectedOutcome` for test methods as mentioned [here](https://nus-cs2103-ay2223s2.github.io/website/schedule/week3/topics.html#c-to-java-junit-junit-basic).

RyanQiu1 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,105 @@
+/**
+ * Represents a Parser to parse date and time.
+ */
+public class Parser {
```
Perhaps the class name could be more specific? As i was confused between a command parser and date/time parser.

lhy-hoyin replied:
`Parser` has been renamed to `DateTimeParser`

RyanQiu1 commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,78 @@
+ * Abstract - cannot be instantiated; serves as template.
+ * Have to be Storable
+ */
+public abstract class Task implements Storable {
```
I like how you this class is done and follows the coding standard 👍

RyanQiu1 commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,63 @@
+/**
+ * Customised how messages are printed to console.
+ */
+public class Ui {
```
Good work on separating the methods and constants according to the java coding standard 👍

RyanQiu1 reviewed (commented):
Overall, i found your code easy to read and understandable, as the project is currently still in progress, i understand some parts of your code are still in work. However, when coding standards is taken into consideration, almost all code comply. Just a few nits to fix.

abenx162 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,277 @@
+        displayTaskCount();
+    }
+
+    void cli() {
```
I feel like this method is too long and has a lot of nesting, perhaps the different commands could be extracted out into separate methods.

lhy-hoyin replied:
Thank you for your feedback! I will make changes accordingly.

abenx162 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,105 @@
+     * TODO: need to catch exceptions
+     *
+     * @param dateStr String representation of a date.
+     * @param seperator Seperator used to separate the day, month, year.
```
should be spelt as "sepArator", not "sepErator".

lhy-hoyin replied:
Thank you for your feedback! I will make changes accordingly.

abenx162 commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,99 @@
+ */
+public class Event extends Task {
+
+    protected LocalDateTime from;
```
could these be more explicitly named? e.g. fromDate or something like that.

lhy-hoyin replied:
Thank you for your feedback! I will make changes accordingly.

abenx162 reviewed (commented):
Overall very well done, I like the extensive commenting that you've done to clarify your code.

lhy-hoyin committed:
Implements a GUI that replaces the function of the console.
GUI offers no additional features. GUI only provides a "imaged" view
to simulate a conversation.

lhy-hoyin committed:
Merge branch 'branch-Level-10'

* branch-Level-10:
  Implements a GUI that replaces the function of the console. GUI offers no additional features. GUI only provides a "imaged" view to simulate a conversation.

lhy-hoyin committed:
Resolve some issues mentioned in PR reviews

lhy-hoyin committed:
Add Javadoc for overridden Applications methods

lhy-hoyin committed:
Add assert statement to document important assumptions

Components of class instance are assumed to exist prior to usage.

Such assumptions may not hold true, since code changes may
unknowingly remove such components.

Assert statements helps to ensure that important assumptions
holds at various points of the code.

Let's add assert statements to ensure that components of the class
instance are not null before interaction with that component.

Using assert statements is preferable since the programmer is
responsible for the validity of the components.

lhy-hoyin committed:
Add assert statements to ensure string validity

Date/Time parser assumes string inputted is not null and contains
some non-whitespace text.

Existing checks to ensure the string validity prior to
passing/calling to parser may unknowing be modified or removed.

Let's add assert statements to ensure the validity of the string
before starting to parse.

Using assert statements here ensure the integrity of the string,
and can prompt for proper user-input checks further up the
abstraction.

lhy-hoyin committed:
Minor code changes

Rename some parameter, method, enum names.
Update some javadocs/comments.
Remove unused imports. Add import to use Flag enum directly.
Refactor duke initialization in main.

lhy-hoyin committed:
Fix bug issue where typing exit (or other keywords) into the GUI
does not trigger the exiting behavior.

lhy-hoyin committed:
Add assert statement to document important assumptions

Components of class instance are assumed to exist prior to usage.

Such assumptions may not hold true, since code changes may
unknowingly remove such components.

Assert statements helps to ensure that important assumptions
holds at various points of the code.

Let's add assert statements to ensure that components of the class
instance are not null before interaction with that component.

Using assert statements is preferable since the programmer is
responsible for the validity of the components.

lhy-hoyin committed:
Abstract Task related classes into a .task sub-package

Task related classes are not categorized into its own sub-package.

Putting the classes into its own sub-package provides better
organization and abstraction.

Let's move task related classes into its own .task sub-package.
Accordingly, update access modifiers of methods of these classes.

This is related to A-MoreOOP and A-Packages in week 3 tasks.

lhy-hoyin committed:
Change toString() of tasks sub-classes to use String.format

lhy-hoyin committed:
Fixed minor bug

lhy-hoyin committed:
Improve code quality by refactoring some code

Code may be generally hard to read or have unsafe practices.

Let's improve code quality through the following refactors:
* Abstract inline method call
* Clarify and improve code comments
* Consolidate duplicate conditional fragments
* Remove unused variables and code blocks
* Rename method name for more clarity
* Replacing magic literals

lhy-hoyin committed:
Add assert statements to ensure string validity

lhy-hoyin committed:
Replace Magic Literal, Add "default" branch, Add javadocs comments.

lhy-hoyin committed:
Spilt Duke.executeCommand into sub-packaged classes

Duke.executeCommand is a very long method (~180 LoC).

A long method makes it very hard to debug, and will become a even
bigger problem as more features are added.

Let's
* extract method
* functionally split the method into various classes

By using classes, additional features would just add more classes,
hence reduce code clutter.

Some other methods were modified to keep the program features
working. Implementation of some methods where improved.

lhy-hoyin committed:
Merge pull request #1 from lhy-hoyin/branch-A-Assertions

Branch-A-Assertions

lhy-hoyin committed:
Merge branch 'master' into branch-A-CodeQuality

# Conflicts:
#	src/main/java/duke/Duke.java
#	src/main/java/duke/Ui.java
#	src/main/java/duke/gui/MainWindow.java
#	src/main/java/duke/storage/Storage.java

lhy-hoyin committed:
Merge pull request #2 from lhy-hoyin/branch-A-CodeQuality

brach-A-CodeQuality

lhy-hoyin committed:
Added java docs, enforced checkstyle

lhy-hoyin committed:
Sets up a custom workflow via GitHub Actions

lhy-hoyin committed:
Rename Parser to DateTimeParser

lhy-hoyin committed:
Add a brief User Guide

lhy-hoyin committed:
no message

lhy-hoyin committed:
no message

lhy-hoyin committed:
UG v0.4 - Change Example display

lhy-hoyin committed:
Bugfix: Remove incorrect assert

Assert to ensure that something is written to file.

We need to consider the situation when user deletes all tasks.
Then task list will be empty (is valid state), meaning there is
nothing to write to file. This falsely fails the assert.

Let's remove the assert.

In general, we should writing a blank to a file.

lhy-hoyin committed:
Fix typo errors, and improve the working of certain parts.

Upgrade UG to version 1.x

lhy-hoyin committed:
Improve header comments to follow format specified in the
coding standard.

lhy-hoyin committed:
Remove unused function; moved TODO remarks out of header comments.

lhy-hoyin committed:
Improve the find function to be not case-sensitive

lhy-hoyin committed:
Add a representative screenshot

lhy-hoyin committed:
Add JUnit tests for Duke

Add Ui Test getRecentMessages() to allow assertEquals in Duke tests

lhy-hoyin committed:
Use a dummy Task instead of Todo in JUnit tests

lhy-hoyin committed:
Add JUnit tests for TaskList. Add and fixed guard clauses.

lhy-hoyin committed:
Allow save file path in Storage to be changed.

lhy-hoyin committed:
Merge branch 'master' into A-MoreTesting

lhy-hoyin committed:
Modify the visibility of getter and setter for Storage saveFilePath

lhy-hoyin committed:
Merge branch 'master' into A-MoreTesting

lhy-hoyin committed:
Make tests write to a different save file.

lhy-hoyin committed:
Update README of repo

lhy-hoyin committed:
Add JUnit tests for Command class.

lhy-hoyin committed:
Merge branch 'A-MoreTesting'

* A-MoreTesting:
  Add JUnit tests for Command class.
  Make tests write to a different save file.
  Add JUnit tests for TaskList. Add and fixed guard clauses.
  Use a dummy Task instead of Todo in JUnit tests
  Add JUnit tests for Duke

JUnit test are not created for individual Command subclasses,
as they rely on Duke to work.
However, Duke is too complex to be mimiced by a stub.
Let's test individual Commands during in integration testing.

lhy-hoyin committed:
Bugfix: Resolved missing GUI response for unrecognized inputs

lhy-hoyin committed:
Handle exceptions due to date/time parsing

Parsing of date/time is not handled and expects user to always
provide date/time in the correct format.

Fatal if not followed. Wrong format will case the program to crash.

Let's finally handle the exception by prompting user to re-type
using the correct format, and not crash the program (fatal).

Also conform to coding styling.

lhy-hoyin committed:
Suppress printing of "add" message during loop-ed testing.

lhy-hoyin committed:
Merge branch 'branch-A-MoreTesting'

* branch-A-MoreTesting:
  Suppress printing of "add" message during loop-ed testing.

lhy-hoyin committed:
Bugfix: GUI overflow will result in clipping of texts

duyLeu commented on src/main/java/duke/command/AddEventCommand.java:
```diff
@@ -0,0 +1,82 @@
+     *                       If start time is empty.
+     *                       If end time is empty.
+     */
+    public AddEventCommand(String userInput) throws DukeException {
+        String taskInfo = userInput.replaceFirst(CMD_KEYWORD, "").trim();
+
+        String[] splitByFrom = taskInfo.split(FROM_KEYWORD, 2);
+        if (splitByFrom.length == 1) {
+            throw new DukeException(String.format("Missing '%s {start time}'.", FROM_KEYWORD));
+        }
+
+        taskDescription = splitByFrom[0].trim();
+        if (taskDescription.isBlank()) {
+            throw new DukeException("Task description cannot be empty.");
+        }
+
+        String[] splitByTo = splitByFrom[1].split(TO_KEYWORD, 2);
+        if (splitByTo.length == 1) {
+            throw new DukeException(String.format("Missing '%s {end time}'.", TO_KEYWORD));
+        }
+
+        fromStr = splitByTo[0].trim();
+        if (fromStr.isBlank()) {
+            throw new DukeException("Start time cannot be empty.");
+        }
+
+        toStr = splitByTo[1].trim();
+        if (toStr.isBlank()) {
+            throw new DukeException("End time cannot be empty.");
+        }
+
+    }
```
This constructor is too long and handles both initialising and parsing the string. 
Maybe you should separate them out into a separate parse function.
This is applicable to all classes in this command package.
